src/github.com/getlantern/detour/detour.go

```diff
@@ -43,325 +38,326 @@ import (
+			newConn.Write(sentBytes, dc.chWrite)
+			newConn.FirstRead(b, dc.chRead)
+			count++
+			// add new connection to connections
+			dc.conns <- newConn
+		case result := <-dc.chRead:
+			conn, n, err := result.conn, result.n, result.err
+			if err != nil {
+				log.Tracef("Read from %s connection to %s failed, closing: %s", typeOf(conn), dc.addr, err)
+				_ = conn.Close()
+				// skip failed connection as we have more
+				if count > 1 {
+					continue
+				}
+				switch conn.ConnType() {
+				case connTypeDirect:
+					// if we haven't dial detour yet, do so now
+					select {
+					case dc.chDialDetourNow <- true:
+						count++
+					default:
+					}
+					continue
+				case connTypeDetour:
+					log.Tracef("Detour connection to %s failed, removing from whitelist", dc.addr)
+					RemoveFromWl(dc.addr)
+					// no more connections, return directly to avoid dead lock
+					return n, err
+				}
 			}
+			log.Tracef("Read %d bytes from %s connection to %s", n, typeOf(conn), dc.addr)
+			dc.incReadBytes(n)
+			return n, err
 		}
-		return
-	}
-	// Hijacked content is usualy encapsulated in one IP packet,
-	// so just check it in one read rather than consecutive reads.
-	if detector.FakeResponse(b) {
-		log.Tracef("Read %d bytes from %s %s, response is hijacked, detour", n, dc.addr, dc.stateDesc())
-		return dc.detour(b)
 	}
-	log.Tracef("Read %d bytes from %s %s, set state to direct", n, dc.addr, dc.stateDesc())
-	dc.setState(stateDirect)
 	return
 }
 
 // followUpRead is called by Read() if a connection's state already settled
-func (dc *Conn) followUpRead(b []byte) (n int, err error) {
-	detector := blockDetector.Load().(*Detector)
-	if n, err = dc.countedRead(b); err != nil {
-		if err == io.EOF {
-			log.Tracef("Read %d bytes from %s %s, EOF", n, dc.addr, dc.stateDesc())
-			return
-		}
-		log.Tracef("Read from %s %s failed: %s", dc.addr, dc.stateDesc(), err)
-		switch {
-		case dc.inState(stateDirect) && detector.TamperingSuspected(err):
-			// to prevent a slow or unstable site from been treated as blocked,
-			// we only check first 4K bytes, which roughly equals to the payload of 3 full packets on Ethernet
-			if atomic.LoadInt64(&dc.readBytes) <= 4096 {
-				log.Tracef("Seems %s still blocked, add to whitelist so will try detour next time", dc.addr)
-				AddToWl(dc.addr, false)
-			}
-		case dc.inState(stateDetour) && wlTemporarily(dc.addr):
-			log.Tracef("Detoured route is not reliable for %s, not whitelist it", dc.addr)
-			RemoveFromWl(dc.addr)
-		}
-		return
-	}
-	// Hijacked content is usualy encapsulated in one IP packet,
-	// so just check it in one read rather than consecutive reads.
-	if dc.inState(stateDirect) && detector.FakeResponse(b) {
-		log.Tracef("%s still content hijacked, add to whitelist so will try detour next time", dc.addr)
-		AddToWl(dc.addr, false)
-		return
+func (dc *Conn) followupRead(b []byte) (n int, err error) {
+	if !dc.withValidConn(func(c conn) { c.FollowupRead(b, dc.chRead) }) {
```

Comment:
Both direct and detour connection can be available at this point. If we don't want to read/write on both connections, we should keep one and close another at first read. It's definitely a bug.